packages/concerto-core/lib/introspect/mapvaluetype.js

```diff
@@ -0,0 +1,127 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const Decorated = require('./decorated');
+const IllegalModelException = require('./illegalmodelexception');
+
+// Types needed for TypeScript generation.
+/* eslint-disable no-unused-vars */
+/* istanbul ignore next */
+if (global === undefined) {
+    const ModelFile = require('./modelfile');
+}
+
+/**
+ * MapValueType defines a Value type of MapDeclaration.
+ *
+ * @extends Decorated
+ * @see See {@link Decorated}
+ * @class
+ * @memberof module:concerto-core
+ */
+class MapValueType extends Decorated {
+    /**
+     * Create an MapValueType.
+     * @param {MapDeclaration} parent - The owner of this property
+     * @param {Object} ast - The AST created by the parser
+     * @param {ModelFile} modelFile - the ModelFile for the Map class
+     * @throws {IllegalModelException}
+     */
+    constructor(parent, ast) {
+        super(ast);
+        this.parent = parent;
+        this.type = null;
+        this.process();
+    }
+
+    /**
+     * Process the AST and build the model
+     *
+     * @throws {IllegalModelException}
+     * @private
+     */
+    process() {
+        super.process();
+        this.type = this.ast.name;
+    }
+
+    /**
+     * Semantic validation of the structure of this class.
+     *
+     * @throws {IllegalModelException}
+     * @protected
+     */
+    validate() {
+        const declarations = this.getModelFile().getAllDeclarations();
+
+        const value = declarations.find(decl => decl.name === this.type);
+
+        if (!value?.isConcept?.()           &&
```

Comment:
Same comment as the other validate call. I'd rewrite this with a positive switch/case rather than all these negative tests.

Reply:
Because all Concerto concepts and primitives are legal MapValueTypes I've since changed this to return true without any checking, but on consideration of your comment  I'm not so sure this is actually correct.

For example, whats to stop a bad AST being fed to the the constructor:

```
{
    '$class': 'concerto.metamodel@1.0.0.MapKeyType',
    name: 'DateTime'
},
{
    '$class': 'concerto.metamodel@1.0.0.AggregateValueType',
    name: 'String3'
}
```

`String3` is not a valid concerto type, but would still pass the current validation step. So, validation is mandatory and should be explicit so as to catch issues early on in the case where a malformed AST is in use. Will add additional validation logic based on your suggestion.